Add location model and serializer

aleksandr-hilko opened the pull request:
In this PR:
- Added Location model which contains 2 fields: address(str) and point(geo field) and has M:1 relation with Trip model. New model was introduced to support work with trips via string view addresses (previously it was possible to feed API only with geo coordinates ). 
So in general now you can create/find trip using human readable addresses along with old coordinates approach. 
 It is possible thanks to the new added geopy library. I'm using geolocation service Nominatim, which abstracts the OpenStreetMap service’s API. 
- Added Location serializer derived from rest_framework_gis.serializers.GeoFeatureModelSerializer. I choose using 'rest_framework_gis' library because it  outputs data in a format that is GeoJSON compatible which is sort of standard now.
- Fixed old tests that failed due to model changes and added some new.

aleksandr-hilko committed:
add Location model and LocationSerializer

Location model that has address and point attributes was added.
Replaced  geocordinates type of fields to use rest_framework_gis library

aleksandr-hilko committed:
fix tests

aleksandr-hilko committed:
add validation tests

aleksandr-hilko committed:
add geopy to requirements

Meathookseed commented on trip_app/core/utils.py:
```diff
@@ -25,12 +30,29 @@ def str_to_geopoint(data):
         E.g. '23.4 23.5', '[23.4 23.5]', '(23.4 23.5)' would be converted into
         the same object Point(23.4, 23.5) """
     lon, lat = re.findall(r"[-+]?\d*\.?\d+|\d+", data)
```
If `re.findall` does not find anything, it returns empty `[]`. So if you try to get `lon` and `lat` it will throw ValueError.

aleksandr-hilko replied:
Wow. Good point, thanks. Should be updated now.
edit: I decided to handle this case with raising the ValidationError error with proper message, also renamed this function to point out that exception can be raised.

Meathookseed commented on trip_app/core/utils.py:
```diff
@@ -25,12 +30,29 @@ def str_to_geopoint(data):
+
+def validate_geo_point(lat, lon):
+    """ Raise validation error on invalid coords """
+    if not -180 <= float(lat) <= 180:
```
I think you can use `abs` method

aleksandr-hilko replied:
Updated to using `abs`.

aleksandr-hilko committed:
add exception handler to str_to_geopoint function

Meathookseed commented on trip_app/core/utils.py:
```diff
@@ -19,18 +23,40 @@ def raise_for_status(resp):
 
 
-def str_to_geopoint(data):
+def str_to_geopoint_or_validation_error(data):
```
I think previous name sounds better

andrey-bulygin replied:
Yeah, you can specify the error case in the docstring (`:raises | Raises` depending on the docstring style). The short name makes more sense.

aleksandr-hilko replied:
I think that it is rather important information to include it in a method name, though it seems rather cumbersome it can make a life easier to the client code, and save time on debugging.
But perhaps it is just my passion to too long explicit names... I will change it.

aleksandr-hilko replied:
Thanks guys! This should be updated now.

andrey-bulygin replied:
You can't place every meaning into the name, so it's better to stick with just the key meaning. Docstrings matter a lot though.

Meathookseed commented on trip_app/core/utils.py:
```diff
@@ -19,18 +23,40 @@ def raise_for_status(resp):
+        the same object Point(23.4, 23.5). """
+    try:
+        lat_str, lon_str = re.findall(r"[-+]?\d*\.?\d+|\d+", data)
+        lat, lon = float(lat_str), float(lon_str)
```
I think its better to make like this:
```
latlon = re.findall(r"[-+]?\d*\.?\d+|\d+", data)
if not latlon:
    raise ValidationError(f"Cannot parse {data} into geo coordinates")
validate_geo_point(*latlon)
return Point(*latlon, srid=4326)
```

aleksandr-hilko replied:
Yep that make sense as well, the reason why I'd preferred catching exception is because converting to float potentially can raise ValidationError as well, though regex should return either empty array or valid numbers, however I want to ensure more in this.

Meathookseed replied:
Casting to float can not throw an error because re.findall would return you array of digits in string representation or empty array.

aleksandr-hilko replied:
Your statement is correct until this re is correct ;)

aleksandr-hilko replied:
@Meathookseed Updated to your version. Thanks.

Meathookseed commented on trip_app/core/utils.py:
```diff
@@ -19,18 +23,40 @@ def raise_for_status(resp):
-    if not -180 <= float(lat) <= 180:
+
+
+def geocode_or_validation_error(address):
```
Here the same with naming.

aleksandr-hilko replied:
Renamed.

andrey-bulygin commented on trip_app/trips/views.py:
```diff
@@ -34,24 +38,40 @@ def get_queryset(self):
+                    geo_point2 = geocode_or_validation_error(address2)
+            except ValidationError:
+                # return empty queryset
+                return Trip.objects.none()
```
This case feels more like a `400` error to me, rather than `200` with empty list. Any specific reasoning on your choice?

aleksandr-hilko replied:
Nothing specific) . I just didn't know that DRF is smart enough to interpret ValidationError as 400 BadResponse, and didn't know how to return the proper response from here ) It look like I haven't to catch the exception anywore, wow. Thanks!

aleksandr-hilko replied:
This should be updated now.

aleksandr-hilko committed:
rename functions

aleksandr-hilko committed:
Modify Dockerfile and docker-compose file

Added install dockerize directives to Dockerfile and added condition to wait for db in docker-compose

aleksandr-hilko committed:
adjusted 'str_to_geopoint' function

Meathookseed commented on trip_app/core/utils.py:
```diff
@@ -21,16 +25,41 @@ def raise_for_status(resp):
+
+        :raises: ValidationError: If incoming data can't be converted
+        to float or if it doesn't fit into coordinates range """
+    latlon = re.findall(r"[-+]?\d*\.?\d+|\d+", data)
```
What if somebody send only 1 value, you should ensure that there is 2 values to avoid IndexError on unpack

aleksandr-hilko replied:
I will check for the length to be 2 otherwise raise Validation error.

aleksandr-hilko replied:
Should be updated now. Thanks.

Meathookseed commented on trip_app/trips/views.py:
```diff
@@ -34,22 +34,33 @@ def get_queryset(self):
+        point2 = query_params.get("point2")
+        address1 = query_params.get("addr1")
+        address2 = query_params.get("addr2")
+        geo_coords = bool(point1 and point2)
```
Here you can use builtin func `all`

aleksandr-hilko replied:
Should be updated now. Thanks.

Meathookseed commented on trip_app/trips/views.py:
```diff
@@ -34,22 +34,33 @@ def get_queryset(self):
+        geo_coords = bool(point1 and point2)
+        address_coords = bool(address1 and address2)
+        if geo_coords or address_coords:
+            if geo_coords:
```
As for me its better to deal with things like this by some factory function, let it be F.E. `get_location_converter` function.
This is not super important, but please take a look at such pattern https://realpython.com/factory-method-python/ and try to use it here

aleksandr-hilko replied:
It is interesting idea, though I haven't found an easy way to apply this pattern here and placed a todo to consider this more in the future.

aleksandr-hilko committed:
small fixes

Meathookseed reviewed (approved):
Sorry for delay
